### PR TITLE
Allow DnsEndpointGroups to set a negativeTtl

### LINKED ISSUE #3938: Armeria does not set a negativeTtl or allow a negativeTtl to be configured for the client
The Netty DNS resolvers do not cache negative responses by default, which can cause typos (or other dns related outages/mishaps) to make armeria overwhelm DNS infrastructure by flooding it with requests to which it gets a negative response.

negativeTtl is a config you can pass to the netty DnsNameResolverBuilder:
https://netty.io/4.1/api/io/netty/resolver/dns/DnsNameResolverBuilder.html#negativeTtl-int-

### PR DESCRIPTION
**Pull request by AngerM**
Motivation:

Fixes #3938 

Modifications:

- Wire up a negativeTtl option to the builders and classes

Result:

- Closes #3938

### CODE REVIEW

**Comment by minwoox** [comment]
Could you also a test case for this?
You might want to refer to the `negativeTtl` test of `RefreshingAddressResolver`:
https://github.com/line/armeria/blob/master/core/src/test/java/com/linecorp/armeria/client/RefreshingAddressResolverTest.java#L258

File: core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroupBuilder.java [hunk 1]
```diff
@@ -82,6 +82,11 @@ public DnsAddressEndpointGroupBuilder ttl(int minTtl, int maxTtl) {
         return (DnsAddressEndpointGroupBuilder) super.ttl(minTtl, maxTtl);
     }
 
+    @Override
+    public DnsEndpointGroupBuilder negativeTtl(int ttl) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroupBuilder.java:86]:
```suggestion
    public DnsAddressEndpointGroupBuilder negativeTtl(int ttl) {
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsServiceEndpointGroupBuilder.java [hunk 2]
```diff
@@ -84,4 +84,9 @@ public DnsServiceEndpointGroupBuilder backoff(Backoff backoff) {
     public DnsServiceEndpointGroupBuilder selectionStrategy(EndpointSelectionStrategy selectionStrategy) {
         return (DnsServiceEndpointGroupBuilder) super.selectionStrategy(selectionStrategy);
     }
+
+    @Override
+    public DnsServiceEndpointGroupBuilder negativeTtl(int ttl) {
+        return (DnsServiceEndpointGroupBuilder) super.negativeTtl(ttl);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsServiceEndpointGroupBuilder.java:90]:
```suggestion
    public DnsServiceEndpointGroupBuilder negativeTtl(int negativeTtl) {
        return (DnsServiceEndpointGroupBuilder) super.negativeTtl(negativeTtl);
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsTextEndpointGroupBuilder.java [hunk 3]
```diff
@@ -91,4 +91,9 @@ public DnsTextEndpointGroupBuilder backoff(Backoff backoff) {
     public DnsTextEndpointGroupBuilder selectionStrategy(EndpointSelectionStrategy selectionStrategy) {
         return (DnsTextEndpointGroupBuilder) super.selectionStrategy(selectionStrategy);
     }
+
+    @Override
+    public DnsTextEndpointGroupBuilder negativeTtl(int ttl) {
+        return (DnsTextEndpointGroupBuilder) super.negativeTtl(ttl);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsTextEndpointGroupBuilder.java:97]:
```suggestion
    public DnsTextEndpointGroupBuilder negativeTtl(int negativeTtl) {
        return (DnsTextEndpointGroupBuilder) super.negativeTtl(negativeTtl);
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroupBuilder.java [hunk 4]
```diff
@@ -82,6 +82,11 @@ public DnsAddressEndpointGroupBuilder ttl(int minTtl, int maxTtl) {
         return (DnsAddressEndpointGroupBuilder) super.ttl(minTtl, maxTtl);
     }
 
+    @Override
+    public DnsAddressEndpointGroupBuilder negativeTtl(int ttl) {
+        return (DnsAddressEndpointGroupBuilder) super.negativeTtl(ttl);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroupBuilder.java:87]:
```suggestion
    public DnsAddressEndpointGroupBuilder negativeTtl(int negativeTtl) {
        return (DnsAddressEndpointGroupBuilder) super.negativeTtl(negativeTtl);
```

**Review by ikhoon — APPROVED** [review]
Thanks, @AngerM!

**Comment by AngerM** [comment]
@ikhoon @trustin (because I see your name in the netty code related to this stuff), having trouble testing this. I exposed the resolver as visible for testing, but after creating a DnsAddressEndpointGroup with an invalid hostname, the resolveCache still seemingly has no entries. Based upon the code and the tests I saw in netty, I was expecting a cached entry for the invalid hostname

**Comment by AngerM** [comment]
When debugging the test I can see the `negativeTtl` makes it way down to the resolverCache, but I just don't see any entries getting added

**Comment by ikhoon** [comment]
Checked the Netty code about DNS cache. If `DnsQuestion` is used for DNS query, results are not cached by `resolveCache`.
This might be not a problem for successful results that are cached in `EndpointGroup`.
https://github.com/line/armeria/blob/766812a3a0c25346c376f0dbb98511e77a7159fd/core/src/main/java/com/linecorp/armeria/internal/client/DefaultDnsNameResolver.java#L76-L85
https://github.com/netty/netty/blob/944a3e5be37bf634e4985075103cd45e618065ab/resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java#L849-L854
https://github.com/netty/netty/blob/23c0bbb9045fda365e906504b559a11bad717faa/resolver-dns/src/main/java/io/netty/resolver/dns/DnsRecordResolveContext.java#L82-L86

To support negative TTL, there are two possible workarounds on Armeria's side.
1. `RefreshingAddressResolver` caches the resolved result by itself. We can use `RefreshingAddressResolver` for `DnsEndpointGroup`.
2. Manually caches the cacheable failure - `UnknownHostException`.

What do you think? @minwoox

**Comment by ikhoon** [comment]
On second thought, as `DnsEndpointGroup` resolves DNS record periodically, `negativeTtl` makes no sense.
If a DNS query fails, the next query will be delayed by a `Backoff` that has almost the same effect as `negativeTtl`.
If you need a constant `negativeTtl`, `Backoff.fixed()` might be suitable for that.

The following line seems to be removed for clarity.
https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroup.java#L96

**Comment by ikhoon** [comment]
On third thought(😅), I think we can clear the cache `Endpoint`s in `EndpointGroup` and make it an empty `EndpointGroup`
1) if a `negativeTtl` is specified and 2) the cause of a DNS failure is `UnknownHostExcepiton`.
Afterwards, the next DNS query will be scheduled after `negativeTtl`.

**Comment by minwoox** [comment]
Yeah, I agree with @ikhoon. We have `Backoff` for failure cases so we do not need `negativeTtl` from the first. 😅 
@AngerM I'm really sorry about this. 😅

**Comment by AngerM** [comment]
@minwoox @ikhoon there is one case where the negative ttl could still have effects: if you have multiple search domains. The way this all came about was that (for legacy reasons) we have multiple search domains configured in our cluster, most of these are going to fail for a given endpoint group, but we still see the slam of netty searching for all the various names when it refreshes

**Comment by minwoox** [comment]
> if you have multiple search domains. The way this all came about was that (for legacy reasons) we have multiple search domains configured in our cluster, most of these are going to fail for a given endpoint group, but we still see the slam of netty searching for all the various names when it refreshes

IIRC, it has nothing to do with the `negativeTtl`.
The DNS resolver appends the search domains to the DNS and sends it DNS servers to find out it it's a valid domain or not.

So, even if we configure the `negativeTtl`, which is basically the same using `Backoff.ofFixed(negativeTtl)`, the process will be the same.

How about configuring the `searchDomains` by yourself if you feel the DNS resolver sends too many queries by appending `searchDomains`?
Of course, we don't have a setter for that so we should provide it via DNS endpoint group builder.

**Comment by AngerM** [comment]
The hope was setting a (longer) negativeTtl so when the resolver appends the search domains, the 'bad' ones hit the cache most of the time with no result, and just the 'good' one end ups getting re-resolved.

**Comment by minwoox** [comment]
> The hope was setting a (longer) negativeTtl so when the resolver appends the search domains, the 'bad' ones hit the cache most of the time with no result, and just the 'good' one end ups getting re-resolved.

Now I got what you meant.
But Netty doesn't provide caching for individual queries that append search domaing.
https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java#L1051

I guess we need to file an issue to the NETTY. 😅

**Comment by minwoox** [comment]
Oops, I found out that it's not cached anyway in Netty. 
https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DnsRecordResolveContext.java#L83-L86

In order to support caching for individual queries, we need to extract [this part](https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java#L212-L251) from Netty and implement it In Armeria's  `RefreshingAddressResolver`.
Then, we can use `RefreshingAddressResolver` for both `WebClient` and `DnsEndpointGroup`.

This is going to be pretty much a big change but I think it's worth trying. Are you still interested in it? 😄 
Just let us know if you don't have time for it. We might implement it later. But before that setting the small number of search domains is the possible workaround I guess.

**Comment by AngerM** [comment]
Hmm. I don't know if I will have time to continue on this, but I will point some of my coworkers here in the hopes one of them may be able to pick it up.

**Comment by minwoox** [comment]
Please, don't worry. 😄 
We can also implement this feature later when we have some time. Just let us know if you really need this feature then we might prioritize it. 😄

**Pull request closed without merging**